### ISSUE TITLE
Tree invalidation after flush

### ISSUE THREAD

**Issue by Guilherme-Interlandi** [libspatialindex/libspatialindex#89]
Hi. I've been using the branch "master" on my project and it worked fine up to now doing only one indexing and then multiple loadings but now I need to index more points dynamically and guarantee that it's saved on disk. I'm trying to use the "flush" function of the RTree with no success, if I print (<<) the tree before flushing, the statistics are fine, but after that, utilization is "0%" and the number of nodes in level is a bizarre number. I could not make a program simple enough to reproduce the problem exactly but the following code shows that apparently there's no data on the tree after the flush.

```
#include <spatialindex/SpatialIndex.h>
#include <iostream>
using namespace std;

using namespace SpatialIndex;
using namespace SpatialIndex::RTree;
using namespace SpatialIndex::StorageManager;

int main(){
        string name("test");
	auto disk = createNewDiskStorageManager(name, 4096);
	id_type id;
	auto tree = createNewRTree(*disk, 0.7, 1000000, 100, 2,
			RTree::RTreeVariant::RV_RSTAR, id);
	for (int i = 0; i < 130; i++) {
		double pt[] = { i, i };
		SpatialIndex::IShape* shape = new SpatialIndex::Point(pt, 2);
		tree->insertData(0, 0, *shape, 0);
	}

	cout << "flush" << endl;
	cout << *tree << endl;
	disk->flush();
	tree->flush();
	cout << "after flush" << endl;
	cout << *tree << endl;
}
```

The output here is:
> flush
> Dimension: 2
> Fill factor: 0.7
> Index capacity: 1000000
> Leaf capacity: 100
> Tight MBRs: enabled
> Near minimum overlap factor: 32
> Reinsert factor: 0.3
> Split distribution factor: 0.4
> Utilization: 65%
> Reads: 188
> Writes: 162
> Hits: 0
> Misses: 0
> Tree height: 2
> Number of data: 130
> Number of nodes: 3
> Level 0 pages: 2
> Level 1 pages: 1
> Splits: 1
> Adjustments: 29
> Query results: 0
> 
> after flush
> Dimension: 2
> Fill factor: 0.7
> Index capacity: 1000000
> Leaf capacity: 100
> Tight MBRs: enabled
> Near minimum overlap factor: 32
> Reinsert factor: 0.3
> Split distribution factor: 0.4

And here is an output from my application:

> flush
> Dimension: 2
> Fill factor: 0.7
> Index capacity: 1000000
> Leaf capacity: 100
> Tight MBRs: enabled
> Near minimum overlap factor: 32
> Reinsert factor: 0.3
> Split distribution factor: 0.4
> Utilization: 29%
> Reads: 30
> Writes: 30
> Hits: 0
> Misses: 0
> Tree height: 1
> Number of data: 29
> Number of nodes: 1
> Level 0 pages: 1
> Splits: 0
> Adjustments: 0
> Query results: 0
> 
> after flush
> Dimension: 2
> Fill factor: 0.7
> Index capacity: 1000000
> Leaf capacity: 100
> Tight MBRs: enabled
> Near minimum overlap factor: 32
> Reinsert factor: 0.3
> Split distribution factor: 0.4
> Utilization: 0%
> Reads: 30
> Writes: 30
> Hits: 0
> Misses: 0
> Tree height: 1
> Number of data: 29
> Number of nodes: 1
> Level 0 pages: 42195368
> Splits: 0
> Adjustments: 0
> Query results: 0

**Comment by Guilherme-Interlandi** [comment]
I added the flush to the Rtree to store the header, on the version 1.8.5, and worked like a charm. Should I make a pull request?

**Comment by hobu** [comment]
Yes we would be interested in the pull request, but I'm not sure if there is a bug or your usage is a bit unexpected or peculiar.

**Comment by Guilherme-Interlandi** [comment]
Did you try to run the code I provided above? It's simple and reproduce the problem. Anyway, I looked at the code and the flush is very simple and only reads variables, do not write anything, so I have no idea why this is happening. Considering that it's an old version and I already have the new one working as I need, I no longer have personal interest in this "bug". I'll make a pull request soon.

**Comment by hobu** [comment]
>  I already have the new one working as I need

The newest libspatialindex does not exhibit this issue?

**Comment by Guilherme-Interlandi** [comment]
It doesn't have the flush on the trees natively, but if you add it exactly as it is implemented on 'master', it works as it should, flushes to the disk and do not corrupt the current tree.
